src/main/scala/org/neo4j/spark/dataframe/Neo4jDataFrame.scala

```diff
@@ -24,67 +23,78 @@ object Neo4jDataFrame {
+          .foreach(chunk => {
+            val params: AnyRef = chunk.map(mapFun).asJava
+            execute(session, statement, Map("rows" -> params).asJava, write = true)
+          })
+      } finally {
+        close(driver, session)
+      }
     })
   }
 
-
-  def createNodes(sc: SparkContext, dataFrame: DataFrame, nodes: (String, Seq[String]), renamedColumns: Map[String, String] = Map.empty): Unit = {
-    dataFrame.cache()
+  def createNodes(sc: SparkContext,
+                  dataFrame: DataFrame,
+                  nodes: (String, Seq[String]),
+                  renamedColumns: Map[String, String] = Map.empty,
+                  partitions: Int = 1,
+                  unwindBatchSize: Int = 10000,
+                  merge: Boolean = false): Unit = {
     val nodeLabel: String = renamedColumns.getOrElse(nodes._2.head, nodes._2.head)
-    val createStatement =
-      s"""
-        UNWIND $$rows as row
-        CREATE (node:${nodes._1.quote} {${nodeLabel.quote} : row.source.${nodeLabel.quote}})
-        SET node = row.node_properties
-        """
-    val partitions = Math.max(1, (dataFrame.count() / 10000).asInstanceOf[Int])
-    val config = Neo4jConfig(sc.getConf)
-    dataFrame.repartition(partitions).foreachPartition(rows => {
-      val params: AnyRef = rows.map(r =>
-        Map(
-          "node_properties" -> nodes._2.map(c => (renamedColumns.getOrElse(c, c), toJava(r.getAs(c)))).toMap.asJava)
-          .asJava).asJava
-      Neo4jDataFrame.execute(config, createStatement, Map("rows" -> params).asJava, write = true)
-    })
+    val createStatement = s"""|
+       |UNWIND $$rows as row
+       |${if (merge) "MERGE" else "CREATE"} (node:${nodes._1.quote} {${nodeLabel.quote} : row.node_properties.${nodeLabel.quote}})
+       |SET node += row.node_properties
+       |""".stripMargin
+    execute(sc, dataFrame, partitions, unwindBatchSize, createStatement, (r: Row) => Map(
+      "node_properties" -> nodes._2.map(c => (renamedColumns.getOrElse(c, c), toJava(r.getAs(c)))).toMap.asJava).asJava
+    )
   }
 
-  def execute(config: Neo4jConfig, query: String, parameters: java.util.Map[String, AnyRef], write: Boolean = false): ResultSummary = {
-    val driver: Driver = config.driver()
-    val session = driver.session()
-    try {
-      val runner = new TransactionWork[ResultSummary]() {
-        override def execute(tx: Transaction): ResultSummary =
-          tx.run(query, parameters).consume()
-      }
-      if (write) {
-        session.writeTransaction(runner)
-      }
-      else
-        session.readTransaction(runner)
-    } finally {
-      if (session.isOpen) session.close()
-      driver.close()
+  private def execute(session: Session, query: String, parameters: java.util.Map[String, AnyRef], write: Boolean = false): ResultSummary = {
+    val runner = new TransactionWork[ResultSummary]() {
+      override def execute(tx: Transaction): ResultSummary =
+        tx.run(query, parameters).consume()
+    }
+    if (write) {
```

Comment:
do we still need all this incl. the TransactionWork if we have the defaultTxMode?

Reply:
Just because I want to left the old implementation as much as possible in order to reduce the impact of the new changes.
If you want I can add the same strategy here and in the other parts